libp2p-xtra/src/node.rs

```diff
@@ -0,0 +1,508 @@
+                    let mut stream = node.listen_on(msg.0)?;
+
+                    loop {
+                        let event = stream.try_next().await?.context("Listener closed")?;
+                        let (peer, control, incoming_substreams, worker) = match event {
+                            ListenerEvent::Upgrade { upgrade, .. } => upgrade.await?,
+                            _ => continue,
+                        };
+
+                        this.send_async_safe(NewConnection {
+                            peer,
+                            control,
+                            incoming_substreams,
+                            worker,
+                        })
+                        .await?;
+                    }
+                }
+            },
+            |error| async move {
+                let _ = this
+                    .send(ListenerFailed {
+                        address: listen_address,
+                        error,
+                    })
+                    .await;
+            },
+        );
+    }
+
+    async fn handle(&mut self, msg: OpenSubstream<Single>) -> Result<Substream, Error> {
+        let peer = msg.peer;
+        let protocols = msg.protocols;
+
+        debug_assert!(
+            protocols.len() == 1,
+            "Type-system enforces that we only try to negotiate one protocol"
+        );
+
+        let (protocol, stream) = self.open_substream(peer, protocols.clone()).await?;
+
+        debug_assert!(
+            protocol == protocols[0],
+            "If negotiation is successful, must have selected the only protocol we sent."
+        );
+
+        Ok(stream)
+    }
+
+    async fn handle(
+        &mut self,
+        msg: OpenSubstream<Multiple>,
+    ) -> Result<(&'static str, Substream), Error> {
+        let peer = msg.peer;
+        let protocols = msg.protocols;
+
+        let (protocol, stream) = self.open_substream(peer, protocols).await?;
+
+        Ok((protocol, stream))
+    }
+}
+
+fn verify_unique_handlers<const N: usize>(
+    inbound_substream_handlers: [(&str, Box<dyn StrongMessageChannel<NewInboundSubstream>>); N],
+) -> HashMap<&str, Box<dyn StrongMessageChannel<NewInboundSubstream>>> {
+    let mut map = HashMap::with_capacity(inbound_substream_handlers.len());
+
+    for (protocol, handler) in inbound_substream_handlers {
+        let previous_handler = map.insert(protocol, handler);
+
+        debug_assert!(
+            previous_handler.is_none(),
+            "Duplicate handler declared for protocol {protocol}"
+        );
+    }
+
+    map
+}
+
+#[async_trait]
+impl xtra::Actor for Node {
+    type Stop = ();
+
+    async fn stopped(self) -> Self::Stop {}
+}
+
+#[derive(Debug)]
+struct ListenerFailed {
+    address: Multiaddr,
+    error: anyhow::Error,
+}
+
+#[derive(Debug)]
+struct FailedToConnect {
+    peer: PeerId,
+    error: anyhow::Error,
+}
+
+#[derive(Debug)]
+struct ConnectionFailed {
+    peer: PeerId,
+    error: anyhow::Error,
+}
```

Comment:
would you mind adding some comments here, and possibly improve naming? they are almost the same, but not quite - after careful look at the handler, one seems to be concerned with inflight requests.

Reply:
They are pretty different in fact:

- `FailedToConnect`: We failed to establish a connection.
- `ConnectionFailed`: An existing connection failed.

It is pretty obvious to me. Given that they are internal messages, I wouldn't worry too much but feel free to add comments to it!

Reply:
would you mind if I renamed `s/ConnectionFiled/ExistingConnectionFailed` ? then they wouldn't need comments at all, and would be more self-descriptive. Happy to push a fixup if you agree. :)

Reply:
It is redundant in my opinion, especially in the presence of `FailedToConnect` :man_shrugging:

I'd like to merge it at this point, feel free to send a follow-up :)

Reply:
sure. incoming a follow-up.